File: Python/Copilot/NBodySimulationKDop.py
import numpy as np
from concurrent.futures import ProcessPoolExecutor

G = 6.67430e-11  # Gravitational constant
dt = 1e-3        # Time step
num_bodies = 1000000  # Number of small bodies
theta = 0.3      # Barnes-Hut approximation parameter

class Body:
    def __init__(self, mass, pos, vel):
        self.mass = mass
        self.pos = np.array(pos, dtype=float)
        self.vel = np.array(vel, dtype=float)
        self.force = np.zeros(3, dtype=float)

class Node:
    def __init__(self, center, size):
        self.body = None
        self.center = np.array(center, dtype=float)
        self.size = size
        self.mass = 0.0
        self.com = np.zeros(3, dtype=float)
        self.children = [None] * 8

class NodePool:
    def __init__(self, size):
        self.pool = [Node([0, 0, 0], 0) for _ in range(size)]
        self.index = 0

    def allocate(self, center, size):
        if self.index >= len(self.pool):
            self.pool.append(Node(center, size))
        else:
            self.pool[self.index].center = np.array(center, dtype=float)
            self.pool[self.index].size = size
            self.pool[self.index].mass = 0.0
            self.pool[self.index].com.fill(0)
            self.pool[self.index].body = None
            self.pool[self.index].children = [None] * 8
        self.index += 1
        return self.pool[self.index - 1]

    def reset(self):
        self.index = 0

def initialize_bodies(n):
    bodies = []

    # Central body
    bodies.append(Body(1e30, [0, 0, 0], [0, 0, 0]))

    # Small bodies in circular orbits
    for _ in range(n):
        radius = 1e11 * np.random.rand()
        speed = np.sqrt(G * bodies[0].mass / radius)
        angle = 2 * np.pi * np.random.rand()

        pos = [radius * np.cos(angle), radius * np.sin(angle), 0]
        vel = [-speed * np.sin(angle), speed * np.cos(angle), 0]
        bodies.append(Body(1e24, pos, vel))

    return bodies

def insert_body(node, body, pool):
    if node.body is None and node.mass == 0.0:
        node.body = body
        node.mass = body.mass
        node.com = body.pos.copy()
        return

    if node.body is not None:
        old_body = node.body
        node.body = None
        insert_body(node, old_body, pool)

    node.mass += body.mass
    node.com = (node.com * (node.mass - body.mass) + body.pos * body.mass) / node.mass

    index = 0
    for i in range(3):
        if body.pos[i] > node.center[i]:
            index |= 1 << i

    if node.children[index] is None:
        new_center = node.center.copy()
        for i in range(3):
            new_center[i] += (1 if index & (1 << i) else -1) * node.size / 4
        node.children[index] = pool.allocate(new_center, node.size / 2)

    insert_body(node.children[index], body, pool)

def build_tree(bodies, pool):
    root = pool.allocate([0, 0, 0], 2 * 1e11)
    for body in bodies:
        insert_body(root, body, pool)
    return root

def calculate_force(body, node):
    if node.body is not None and node.body is not body:
        dx = node.body.pos - body.pos
        dist = np.linalg.norm(dx)
        force = G * body.mass * node.body.mass / dist**3 * dx
        body.force += force
        return

    dx = node.com - body.pos
    dist = np.linalg.norm(dx)

    if node.size / dist < theta:
        force = G * body.mass * node.mass / dist**3 * dx
        body.force += force
        return

    for child in node.children:
        if child is not None:
            calculate_force(body, child)

def calculate_forces(bodies, start, end, root):
    for i in range(start, end):
        bodies[i].force.fill(0)
        calculate_force(bodies[i], root)
    return bodies[start:end]

def simulate_step(bodies, pool):
    num_workers = 4  # Adjust based on your system's capabilities
    chunk_size = len(bodies) // num_workers

    pool.reset()
    root = build_tree(bodies, pool)

    with ProcessPoolExecutor(max_workers=num_workers) as executor:
        futures = [executor.submit(calculate_forces, bodies, i, min(i + chunk_size, len(bodies)), root)
                   for i in range(0, len(bodies), chunk_size)]
        results = [future.result() for future in futures]

    for i, result in enumerate(results):
        bodies[i * chunk_size:(i + 1) * chunk_size] = result

    # Update velocities and positions
    for body in bodies:
        body.vel += body.force / body.mass * dt
        body.pos += body.vel * dt

def calculate_energy(bodies):
    kinetic_energy = 0
    potential_energy = 0

    for i, body_i in enumerate(bodies):
        kinetic_energy += 0.5 * body_i.mass * np.dot(body_i.vel, body_i.vel)

        for j in range(i + 1, len(bodies)):
            body_j = bodies[j]
            dx = body_j.pos - body_i.pos
            dist = np.linalg.norm(dx)
            potential_energy -= G * body_i.mass * body_j.mass / dist

    return kinetic_energy + potential_energy

def main():
    np.random.seed(1234)

    # Initialize bodies
    bodies = initialize_bodies(num_bodies)

    # Create a node pool
    pool = NodePool(num_bodies)

    # Calculate initial energy
    initial_energy = calculate_energy(bodies)
    print(f"Initial Energy: {initial_energy:.6e}")

    # Run simulation for 1000 steps
    for _ in range(1000):
        simulate_step(bodies, pool)

    # Calculate final energy
    final_energy = calculate_energy(bodies)
    print(f"Final Energy: {final_energy:.6e}")

if __name__ == "__main__":
    main()

File: Julia/Copilot/NBodySimulationKDop.jl
using Random
using Base.Threads

const G = 6.67430e-11  # Gravitational constant
const dt = 1e-3        # Time step
const numBodies = 1000000  # Number of small bodies
const theta = 0.3      # Barnes-Hut approximation parameter

struct Body
    mass::Float64
    pos::Vector{Float64}
    vel::Vector{Float64}
    force::Vector{Float64}
end

mutable struct Node
    body::Union{Body, Nothing}
    center::Vector{Float64}
    size::Float64
    mass::Float64
    com::Vector{Float64}
    children::Vector{Union{Node, Nothing}}

    Node(center, size) = new(nothing, center, size, 0.0, [0.0, 0.0, 0.0], Vector{Union{Node, Nothing}}(undef, 8))
end

mutable struct NodePool
    pool::Vector{Node}
    index::Int

    NodePool(size::Int) = new(Vector{Node}(undef, size), 1)

    function allocate!(pool::NodePool, center::Vector{Float64}, size::Float64)
        if pool.index > length(pool.pool)
            push!(pool.pool, Node(center, size))
        else
            pool.pool[pool.index] = Node(center, size)
        end
        pool.index += 1
        return pool.pool[pool.index - 1]
    end

    function reset!(pool::NodePool)
        pool.index = 1
    end
end

function initializeBodies(n::Int)
    bodies = Vector{Body}(undef, n + 1)

    # Central body
    bodies[1] = Body(1e30, [0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0])

    # Small bodies in circular orbits
    for i in 2:n+1
        radius = 1e11 * rand()
        speed = sqrt(G * bodies[1].mass / radius)
        angle = 2 * π * rand()

        bodies[i] = Body(1e24,
                         [radius * cos(angle), radius * sin(angle), 0.0],
                         [-speed * sin(angle), speed * cos(angle), 0.0],
                         [0.0, 0.0, 0.0])
    end

    return bodies
end

function buildTree(bodies::Vector{Body}, pool::NodePool)
    root = allocate!(pool, [0.0, 0.0, 0.0], 2 * 1e11)
    @threads for body in bodies
        insertBody!(root, body, pool)
    end
    return root
end

function insertBody!(node::Node, body::Body, pool::NodePool)
    if node.body === nothing && node.mass == 0.0
        node.body = body
        node.mass = body.mass
        node.com = copy(body.pos)
        return
    end

    if node.body !== nothing
        oldBody = node.body
        node.body = nothing
        insertBody!(node, oldBody, pool)
    end

    node.mass += body.mass
    for i in 1:3
        node.com[i] = (node.com[i] * (node.mass - body.mass) + body.pos[i] * body.mass) / node.mass
    end

    index = 0
    for i in 1:3
        if body.pos[i] > node.center[i]
            index |= 1 << (i - 1)
        end
    end

    if node.children[index] === nothing
        newCenter = copy(node.center)
        for i in 1:3
            newCenter[i] += (index & (1 << (i - 1)) != 0 ? 1 : -1) * node.size / 4
        end
        node.children[index] = allocate!(pool, newCenter, node.size / 2)
    end

    insertBody!(node.children[index], body, pool)
end

function calculateForce!(body::Body, node::Node)
    if node.body !== nothing && node.body !== body
        dx = node.body.pos[1] - body.pos[1]
        dy = node.body.pos[2] - body.pos[2]
        dz = node.body.pos[3] - body.pos[3]
        dist = sqrt(dx*dx + dy*dy + dz*dz)
        force = G * body.mass * node.body.mass / (dist * dist * dist)

        body.force[1] += force * dx
        body.force[2] += force * dy
        body.force[3] += force * dz
        return
    end

    dx = node.com[1] - body.pos[1]
    dy = node.com[2] - body.pos[2]
    dz = node.com[3] - body.pos[3]
    dist = sqrt(dx*dx + dy*dy + dz*dz)

    if node.size / dist < theta
        force = G * body.mass * node.mass / (dist * dist * dist)

        body.force[1] += force * dx
        body.force[2] += force * dy
        body.force[3] += force * dz
        return
    end

    for child in node.children
        if child !== nothing
            calculateForce!(body, child)
        end
    end
end

function simulateStep!(bodies::Vector{Body}, pool::NodePool)
    # Reset forces
    @threads for i in 1:length(bodies)
        bodies[i].force .= 0.0
    end

    # Build kD-tree
    reset!(pool)
    root = buildTree(bodies, pool)

    # Calculate forces using kD-tree
    @threads for i in 1:length(bodies)
        calculateForce!(bodies[i], root)
    end

    # Update velocities and positions
    @threads for i in 1:length(bodies)
        bodies[i].vel .+= bodies[i].force ./ bodies[i].mass .* dt
        bodies[i].pos .+= bodies[i].vel .* dt
    end
end

function calculateEnergy(bodies::Vector{Body})
    kineticEnergy = 0.0
    potentialEnergy = 0.0

    @threads for i in 1:length(bodies)
        kineticEnergy += 0.5 * bodies[i].mass * sum(bodies[i].vel .^ 2)

        for j in i+1:length(bodies)
            dx = bodies[j].pos[1] - bodies[i].pos[1]
            dy = bodies[j].pos[2] - bodies[i].pos[2]
            dz = bodies[j].pos[3] - bodies[i].pos[3]
            dist = sqrt(dx*dx + dy*dy + dz*dz)
            potentialEnergy -= G * bodies[i].mass * bodies[j].mass / dist
        end
    end

    return kineticEnergy + potentialEnergy
end

function main()
    Random.seed!(1234)

    # Initialize bodies
    bodies = initializeBodies(numBodies)

    # Create a node pool
    pool = NodePool(numBodies)

    # Calculate initial energy
    initialEnergy = calculateEnergy(bodies)
    println("Initial Energy: $initialEnergy")

    # Run simulation for 1000 steps
    for step in 1:1000
        simulateStep!(bodies, pool)
    end

    # Calculate final energy
    finalEnergy = calculateEnergy(bodies)
    println("Final Energy: $finalEnergy")
end

main()